hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/StakeIdChangeType.java

```diff
@@ -0,0 +1,157 @@
+ * staked to an account, which is staked to a node.
+ */
+public enum StakeIdChangeType {
+    /* ---  Cases ending with staking to a node */
+    FROM_ABSENT_TO_NODE {
+        @Override
+        boolean awardsToNode() {
+            return true;
+        }
+    },
+    FROM_ACCOUNT_TO_NODE {
+        @Override
+        boolean withdrawsFromAccount() {
+            return true;
+        }
+
+        @Override
+        boolean awardsToNode() {
+            return true;
+        }
+    },
+    FROM_NODE_TO_NODE {
+        @Override
+        boolean withdrawsFromNode() {
+            return true;
+        }
+
+        @Override
+        boolean awardsToNode() {
+            return true;
+        }
+    },
+    /* --- Cases ending with staking to an account */
+    FROM_ABSENT_TO_ACCOUNT {
+        @Override
+        boolean awardsToAccount() {
+            return true;
+        }
+    },
+    FROM_NODE_TO_ACCOUNT {
+        @Override
+        boolean withdrawsFromNode() {
+            return true;
+        }
+
+        @Override
+        boolean awardsToAccount() {
+            return true;
+        }
+    },
+    FROM_ACCOUNT_TO_ACCOUNT {
+        @Override
+        boolean withdrawsFromAccount() {
+            return true;
+        }
+
+        @Override
+        boolean awardsToAccount() {
+            return true;
+        }
+    },
+    /* --- Cases ending with absent staking */
+    FROM_ABSENT_TO_ABSENT {},
+    FROM_ACCOUNT_TO_ABSENT {
+        @Override
+        boolean withdrawsFromAccount() {
+            return true;
+        }
+    },
+    FROM_NODE_TO_ABSENT {
+        @Override
+        boolean withdrawsFromNode() {
+            return true;
+        }
+    };
+
+    public static StakeIdChangeType forCase(
+            @Nullable final Account currentAccount, @NonNull final Account modifiedAccount) {
+        final var curStakedIdCase =
+                currentAccount == null ? UNSET : currentAccount.stakedId().kind();
+        final var newStakedIdCase = modifiedAccount.stakedId().kind();
+
+        // Ends with staking to a node
+        if (newStakedIdCase.equals(STAKED_NODE_ID)) {
+            if (curStakedIdCase.equals(UNSET)) {
+                return FROM_ABSENT_TO_NODE;
+            } else if (curStakedIdCase.equals(STAKED_ACCOUNT_ID)) {
+                return FROM_ACCOUNT_TO_NODE;
+            } else {
+                // We don't care if newStakedId == curStakedId, just run the withdraw/reward logic;
+                // it may be necessary in any case if declineReward changed
+                return FROM_NODE_TO_NODE;
+            }
+        } else if (newStakedIdCase.equals(STAKED_ACCOUNT_ID)) {
+            if (curStakedIdCase.equals(UNSET)) {
+                return FROM_ABSENT_TO_ACCOUNT;
+            } else if (curStakedIdCase.equals(STAKED_ACCOUNT_ID)) {
+                return FROM_ACCOUNT_TO_ACCOUNT;
+            } else {
+                return FROM_NODE_TO_ACCOUNT;
+            }
+        } else {
```

Comment:
For readability, could add a comment here,
```
// We are in the newStakedIdCase == UNSET branch now
```

Reply:
Yes done. Thanks